# Hyperparameter Tuning for Cervical Cancer Classification using HyperDrive

In [1]:
from azureml.core import Workspace, Experiment
from azureml.core.compute import ComputeTarget, AmlCompute
from azureml.core.compute_target import ComputeTargetException
from azureml.widgets import RunDetails
from azureml.train.sklearn import SKLearn
from azureml.train.hyperdrive.run import PrimaryMetricGoal
from azureml.train.hyperdrive.policy import BanditPolicy
from azureml.train.hyperdrive.sampling import RandomParameterSampling
from azureml.train.hyperdrive.runconfig import HyperDriveConfig
from azureml.train.hyperdrive.parameter_expressions import uniform, choice
import os

Firstly, we require our experiment to be defined within our workspace, the information about the same maybe observed as follows.

In [2]:
ws = Workspace.from_config()
experiment_name = 'cervical-cancer-classification-hyperdrive-biopsy'

experiment=Experiment(ws, experiment_name)

print('Workspace name: ' + ws.name, 
      'Azure region: ' + ws.location, 
      'Subscription id: ' + ws.subscription_id, 
      'Resource group: ' + ws.resource_group, sep = '\n')

run = experiment.start_logging()

Performing interactive authentication. Please follow the instructions on the terminal.
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code EUPKGFNAK to authenticate.
You have logged in. Now let us find all the subscriptions to which you have access...
Interactive authentication successfully completed.
Workspace name: quick-starts-ws-135758
Azure region: southcentralus
Subscription id: 510b94ba-e453-4417-988b-fbdc37b55ca7
Resource group: aml-quickstarts-135758


## Dataset

We're using the [UCI's Machine Learning repository's Cervical cancer (Risk Factors) Data Set](https://archive.ics.uci.edu/ml/datasets/Cervical+cancer+%28Risk+Factors%29). The dataset was collected at 'Hospital Universitario de Caracas' in Caracas, Venezuela. The dataset comprises demographic information, habits, and historic medical records of 858 patients. Several patients decided not to answer some of the questions because of privacy concerns (missing values).

**Citation: Kelwin Fernandes, Jaime S. Cardoso, and Jessica Fernandes. 'Transfer Learning with Partial Observability Applied to Cervical Cancer Screening.' Iberian Conference on Pattern Recognition and Image Analysis. Springer International Publishing, 2017.**

Our target variable here is **Biopsy**. Biopsy is a sample of tissue taken from the body in order to examine it more closely. A doctor should recommend a biopsy when an initial test suggests an area of tissue in the body isn't normal. Doctors may call an area of abnormal tissue a lesion, a tumor, or a mass.
Here this categorical variable contains the *Biopsy Test Result*.

More information about how the data is cleaned and used maybe found in the `data-cleaning.ipynb` and `train.py` files.

## Hyperdrive Configuration

We firstly need a CPU cluster in order to run our operations on.

In [3]:
# Choose a name for your CPU cluster
cpu_cluster_name = "cpucluster"

# Verify that cluster does not exist already
try:
    cpu_cluster = ComputeTarget(workspace=ws, name=cpu_cluster_name)
    print('Found existing cluster, use it.')
except ComputeTargetException:
    compute_config = AmlCompute.provisioning_configuration(vm_size='STANDARD_D2_V2',
                                                           max_nodes=4)
    cpu_cluster = ComputeTarget.create(ws, cpu_cluster_name, compute_config)

cpu_cluster.wait_for_completion(show_output=True)


Creating
Succeeded
AmlCompute wait for completion finished

Minimum number of nodes requested have been provisioned


Being a classification problem, I'll be using **Logistic Regression** (read `train.py` for greater insight) whose hyperparameters are tuned using the following configuration.

**Bandit Policy** <br/>
Bandit policy is based on slack factor/slack amount and evaluation interval. Bandit terminates runs where the primary metric is not within the specified slack factor/slack amount compared to the best performing run. Unlike Truncation policy it doesn't calculate primary metric for all runs only to delete a percentage of them, but termminate it as soon as the primary metric doesn't satisfy slack amount, omitting unnecessary baggage. It also omits the need to calculate running Median, making it less computationally cumbersome unlike MedianStoppingPolicy.

**Random Parameter Sampling** <br/>
Random sampling supports discrete and continuous hyperparameters. In random sampling, hyperparameter values are randomly selected from the defined search space. It supports early termination of low-performance runs. Unlike other methods, this gives us a wide exploratory range, which is good to do when we don't have much idea about the parameters. It can also be used do an initial search with random sampling and then refine the search space to improve results.

Since the data isn't quite balanced, **Weighted Average Precision Score** as my primary metric.

In [17]:
early_termination_policy = BanditPolicy(evaluation_interval=1, slack_factor=0.001)

param_sampling = RandomParameterSampling(
    {
        "--C":uniform(0.4,0.5),
        "--max_iter": choice(1000,1100,1200,1300,1400)
    }
)

if "training" not in os.listdir():
    os.mkdir("./training")

estimator = SKLearn(source_directory='./', 
                entry_script='train.py', compute_target=cpu_cluster)

hyperdrive_run_config = HyperDriveConfig(estimator=estimator,
                                hyperparameter_sampling=param_sampling,
                                policy=early_termination_policy,
                                primary_metric_name='average_precision_score_weighted',
                                primary_metric_goal=PrimaryMetricGoal.MAXIMIZE,
                                max_total_runs=20,
                                max_concurrent_runs=4)

In [18]:
#TODO: Submit your experiment
hyperdrive_run = experiment.submit(config=hyperdrive_run_config)

## Run Details

The `Rundetails` widget, as the name suggests gives us greater insight about how the Run is proceeding, enabling us to monitor and understand the situation, thereby dealing with it accordingly.

In [19]:
RunDetails(hyperdrive_run).show()
hyperdrive_run.wait_for_completion(show_output=True)

_HyperDriveWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO'…

RunId: HD_f12cdac0-4ee8-45ff-9147-ad454053ed34
Web View: https://ml.azure.com/experiments/cervical-cancer-classification-hyperdrive-biopsy/runs/HD_f12cdac0-4ee8-45ff-9147-ad454053ed34?wsid=/subscriptions/510b94ba-e453-4417-988b-fbdc37b55ca7/resourcegroups/aml-quickstarts-135758/workspaces/quick-starts-ws-135758

Streaming azureml-logs/hyperdrive.txt

"<START>[2021-01-23T08:34:09.843453][API][INFO]Experiment created<END>\n""<START>[2021-01-23T08:34:10.254627][GENERATOR][INFO]Trying to sample '4' jobs from the hyperparameter space<END>\n""<START>[2021-01-23T08:34:10.428293][GENERATOR][INFO]Successfully sampled '4' jobs, they will soon be submitted to the execution target.<END>\n"<START>[2021-01-23T08:34:11.2114404Z][SCHEDULER][INFO]The execution environment is being prepared. Please be patient as it can take a few minutes.<END>

Execution Summary
RunId: HD_f12cdac0-4ee8-45ff-9147-ad454053ed34
Web View: https://ml.azure.com/experiments/cervical-cancer-classification-hyperdrive-biopsy/runs

{'runId': 'HD_f12cdac0-4ee8-45ff-9147-ad454053ed34',
 'target': 'cpucluster',
 'status': 'Completed',
 'startTimeUtc': '2021-01-23T08:34:09.663931Z',
 'endTimeUtc': '2021-01-23T08:53:42.162361Z',
 'properties': {'primary_metric_config': '{"name": "average_precision_score_weighted", "goal": "maximize"}',
  'resume_from': 'null',
  'runTemplate': 'HyperDrive',
  'azureml.runsource': 'hyperdrive',
  'platform': 'AML',
  'ContentSnapshotId': '99ac7e68-279d-4fd5-a1b2-beed939f7460',
  'score': '0.10382513661202186',
  'best_child_run_id': 'HD_f12cdac0-4ee8-45ff-9147-ad454053ed34_7',
  'best_metric_status': 'Succeeded'},
 'inputDatasets': [],
 'outputDatasets': [],
 'logFiles': {'azureml-logs/hyperdrive.txt': 'https://mlstrg135758.blob.core.windows.net/azureml/ExperimentRun/dcid.HD_f12cdac0-4ee8-45ff-9147-ad454053ed34/azureml-logs/hyperdrive.txt?sv=2019-02-02&sr=b&sig=BuSGje6kl6wQLbb3qdLkgh0V%2FZcn%2BhTjbov265fOQDI%3D&st=2021-01-23T08%3A44%3A08Z&se=2021-01-23T16%3A54%3A08Z&sp=r'}}

## Best Model

We hereby observe the best model to recommend
* **Regularisation Strength** - 0.4862058369948942
* **Maximum Interations** - 1400

Giving us an average weighted precision score of 0.10382513661202186

In [20]:
best_run = hyperdrive_run.get_best_run_by_primary_metric()
best_run_metrics = best_run.get_metrics()
parameter_values = best_run.get_details()['runDefinition']['arguments']

print('Best Run ID : ',best_run.id)
print('\n Metrics: ',best_run_metrics)

print('\n Parameters:', parameter_values)
print('\n Average Weighted Precision Score:', best_run_metrics['average_precision_score_weighted'])

Best Run ID :  HD_f12cdac0-4ee8-45ff-9147-ad454053ed34_7

 Metrics:  {'Regularization Strength:': 0.4862058369948942, 'Max iterations:': 1400, 'average_precision_score_weighted': 0.10382513661202186}

 Parameters: ['--C', '0.4862058369948942', '--max_iter', '1400']

 Average Weighted Precision Score: 0.10382513661202186
